lib/pbench/cli/server/report.py

```diff
@@ -514,6 +515,108 @@ def report_uploads(options: dict[str, Any]):
+    """Report dataset statistics by agent version"""
+
+    v_pattern = re.compile(r"v?(?P<major>\d+\.\d+)(?:\.\d+)?(?:-\w+)")
+
+    @dataclass
+    class Daterange:
+        first: Optional[datetime.datetime] = None
+        last: Optional[datetime.datetime] = None
+
+        def add(self, date: datetime.datetime):
+            if self.first is None or date < self.first:
+                self.first = date
+            if self.last is None or date > self.last:
+                self.last = date
+
+    def print_versions(target: dict[str, Daterange], counts: dict[str, int]):
+        click.echo(f"  {'Count':^10s} {'Version':^22s} {'First':^12s} {'Last':^12s}")
+        click.echo(f"  {'':-<10} {'':-<22} {'':-<12} {'':-<12}")
+        for version, dates in sorted(target.items(), key=lambda k: k[1].last):
+            count = counts[version]
+            first = humanize.naturaldate(dates.first)
+            last = humanize.naturaldate(dates.last)
+            click.echo(f"  {count:>10,d} {version!r:^22s} {first:>12s} {last:>12s}")
+
+    watcher.update("analyzing version patterns")
+    since = options.get("since")
+    until = options.get("until")
+
+    if since and until and since > until:
+        raise Exception("The --until value must be later than the --since value")
+
+    rows = Database.db_session.query(
+        cast(Metadata.value["pbench", "date"].as_string(), TZDateTime).label("date"),
+        Metadata.value["pbench", "rpm-version"].as_string().label("version"),
+    ).filter(Metadata.key == "metalog")
+
+    count = 0
+    dateless = 0
+    versionless = 0
+
+    versions = defaultdict(int)
+    majorversions = defaultdict(int)
+    nonversions = defaultdict(int)
+    range = defaultdict(Daterange)
+    majorrange = defaultdict(Daterange)
+    nonversionrange = defaultdict(Daterange)
+
+    filters = []
+
+    # Create a subquery from our basic select parameters so that we can use
+    # the label (SQL "AS date") in our WHERE filter clauses. (In a direct query
+    # PostgreSQL doesn't allow filtering on renamed columns.)
+    subquery = rows.subquery()
+    query = Database.db_session.query(subquery.c.date, subquery.c.version).order_by(
+        subquery.c.date
+    )
+
+    if since:
+        verifier.status(f"Filter since {since}")
+        filters.append(subquery.c.date >= since)
+    if until:
+        verifier.status(f"Filter until {until}")
+        filters.append(subquery.c.date <= until)
+    if filters:
+        query = query.filter(*filters)
+    rows = query.execution_options(stream_results=True).yield_per(SQL_CHUNK)
+
+    for row in rows:
+        count += 1
+        date: datetime.datetime = row[0]
+        version = row[1]
+        if not isinstance(version, str):
+            versionless += 1
+            continue
+        if not isinstance(date, datetime.datetime):
+            dateless += 1
+            date = datetime.datetime.fromtimestamp(0.0)
+        m = v_pattern.match(version)
+        if m:
```

Comment:
Just to confirm, you really want `match()` and not `fullmatch()` (or `search()`) here?

(I would recommend avoiding `match()` and using either `fullmatch()` or explicit anchoring, to avoid this question.)

Reply:
It doesn't really matter in this case, although I suppose if I used search I could drop the leading `v`.